File: easy/3633_earliest_finish_time_for_land_and_water_ride_i/solution.cpp
#include <vector>

class Solution {
public:
  int earliestFinishTime(std::vector<int> &landStartTime,
                         std::vector<int> &landDuration,
                         std::vector<int> &waterStartTime,
                         std::vector<int> &waterDuration) {
    int earliestLandFinishTime = INT_MAX;
    for (size_t i = 0; i < landStartTime.size(); ++i)
      earliestLandFinishTime =
          std::min(earliestLandFinishTime, landStartTime[i] + landDuration[i]);

    int earliestWaterFinishTime = INT_MAX;
    int earliestFinishTime = INT_MAX;

    for (size_t i = 0; i < waterStartTime.size(); ++i) {
      earliestWaterFinishTime = std::min(earliestWaterFinishTime,
                                         waterStartTime[i] + waterDuration[i]);
      earliestFinishTime =
          std::min(earliestFinishTime,
                   std::max(earliestLandFinishTime, waterStartTime[i]) +
                       waterDuration[i]);
    }

    for (size_t i = 0; i < landStartTime.size(); ++i)
      earliestFinishTime =
          std::min(earliestFinishTime,
                   std::max(earliestWaterFinishTime, landStartTime[i]) +
                       landDuration[i]);

    return earliestFinishTime;
  }
};

File: easy/3633_earliest_finish_time_for_land_and_water_ride_i/solution.py
class Solution:
    def earliestFinishTime(
        self,
        landStartTime: list[int],
        landDuration: list[int],
        waterStartTime: list[int],
        waterDuration: list[int],
    ) -> int:
        earliestLandFinishTime = float("inf")
        for i in range(len(landStartTime)):
            if landStartTime[i] >= earliestLandFinishTime:
                continue

            earliestLandFinishTime = min(
                earliestLandFinishTime, landStartTime[i] + landDuration[i]
            )

        earliestWaterFinishTime = float("inf")
        for i in range(len(waterStartTime)):
            if waterStartTime[i] >= earliestLandFinishTime:
                continue

            earliestWaterFinishTime = min(
                earliestWaterFinishTime, waterStartTime[i] + waterDuration[i]
            )

        earliestFinishTime = float("inf")

        for i in range(len(landStartTime)):
            earliestFinishTime = min(
                earliestFinishTime,
                max(earliestWaterFinishTime, landStartTime[i])
                + landDuration[i],
            )

        for i in range(len(waterStartTime)):
            earliestFinishTime = min(
                earliestFinishTime,
                max(earliestLandFinishTime, waterStartTime[i])
                + waterDuration[i],
            )

        return earliestFinishTime
